src/gradient.cc

```diff
@@ -0,0 +1,335 @@
+/*
+ *
+ * Conky, a system monitor, based on torsmo
+ *
+ * Any original torsmo code is licensed under the BSD license
+ *
+ * All code written since the fork of torsmo is licensed under the GPL
+ *
+ * Please see COPYING for details
+ *
+ * Copyright (c) 2004, Hannu Saransaari and Lauri Hakkarainen
+ * Copyright (c) 2005-2021 Brenden Matthews, Philip Kovacs, et. al.
+ *	(see AUTHORS)
+ * All rights reserved.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+#include "gradient.h"
+#include "conky.h"
+#include "logging.h"
+
+#ifdef BUILD_X11
+#include "x11.h"
+#endif /* BUILD_X11 */
+
+namespace conky {
+bool gradient_factory::is_set = false;
+short gradient_factory::colour_depth = 0;
+long gradient_factory::mask[3];
+short gradient_factory::shift[3];
+
+gradient_factory::gradient_factory(int width, unsigned long first_colour,
+                                   unsigned long last_colour) {
+  // Make sure the width is always at least 2
+  this->width = std::max(2, width);
+  this->first_colour = first_colour;
+  this->last_colour = last_colour;
+
+  if (!is_set) {
+    setup_colour_depth();
+    setup_shifts();
+    setup_masks();
+    is_set = true;
+  }
+}
+
+void gradient_factory::setup_shifts() {
+  shift[0] = (2 * colour_depth / 3 + colour_depth % 3);
+  shift[1] = (colour_depth / 3);
+  shift[2] = 0;
+}
+
+void gradient_factory::setup_masks() {
+  mask[0] = mask[1] = mask[2] = 0;
+
+  for (int i = (colour_depth / 3) - 1; i >= 0; i--) {
+    mask[0] |= 1 << i;
+    mask[1] |= 1 << i;
+    mask[2] |= 1 << i;
+  }
+
+  if (colour_depth % 3 == 1) {
+    mask[1] |= 1 << (colour_depth / 3);
+  }
+
+  for (int i = 0; i < 3; i++) {
+    mask[i] = mask[i] << shift[i];
+  }
+}
+
+void gradient_factory::setup_colour_depth() {
+#ifdef BUILD_X11
+  if (state == nullptr) { // testing purposes
```

Comment:
This is okay for now, but it would be better to initialize the state with appropriate values as part of the setup for the tests.